### PR TITLE
Generate transportKey and write it to the notaryProviderContract. Remove transport encryption based on OTEnvelope.

### PR DESCRIPTION
**Pull request by benma**
This PR puts a `<transportKey>` element into the notaryProviderContract, containing the notary's public transport key. Using [zcert_public_txt](http://czmq.zeromq.org/manual:zcert) returns the [Z85 encoding](http://rfc.zeromq.org/spec:32) of the key. Despite the Z85 RFC explicitly mentioning XML in the reasoning of the encoding, the encoding contains the `<>` characters. This is why I also base64-encode it. @toxeus, @OttoAllmendinger, do you have a better solutions in mind? I did not yet investigate PEM or section markers, I figured @toxeus is the expert on that.

The public and the private key are stored in `server_data/credentials/<serverNymID>/transportKey` (and `_secret`). The notary generates/reads this. The client reads the transport public key from the notary provider contract.

### CODE REVIEW

**Comment by OttoAllmendinger** [comment]
About Z85: I think it's supposed to be "easier" to embed because you only need to escape the `&<>` characters, but you have to do some escaping nonetheless. 

Maybe we should just base64-encode the raw binary key data and skip the Z85 encoding altogether?

**Comment by benma** [comment]
I did that initially, then I saw that `zsock_set_curve_serverkey(sock, key)` (as used before) took the Z85-encoded key, and I did not find an encoding function in czmq after a quick search. So I made my life easy ;) MIght might give this another try, though, because the binary is also shorter.

**Comment by benma** [comment]
Oh, and Twin Peaks is the best show of all time.

File: src/core/OTServerContract.cpp [hunk 1]
```diff
@@ -265,10 +297,23 @@ int32_t OTServerContract::ProcessXMLNode(irr::io::IrrXMLReader*& xml)
...
+    }
+    if (!strcmp("transportKey", xml->getNodeName())) {
+        if (!SkipToTextField(xml)) {
+            return -1;
+        }
+        const char* transportKeyB64 = xml->getNodeData();
+        size_t outLen;
+        m_transportKey = reinterpret_cast<char*>(
+            OTCrypto::It()->Base64Decode(transportKeyB64, &outLen, false));
+        if (outLen != 41) {
```
Comment (mhamrle) [inline src/core/OTServerContract.cpp]: Can we use constant instead of magic number 41?

**Comment by OttoAllmendinger** [comment]
It's probably this one with the `_bin` suffix:

```
CZMQ_EXPORT void zsock_set_curve_serverkey_bin (void *self, const byte *curve_serverkey);
```

> Oh, and Twin Peaks is the best show of all time.

File: src/client/OTServerConnection.cpp [hunk 2]
```diff
@@ -240,7 +239,7 @@ void OTServerConnection::send(OTServerContract* pServerContract, Nym* pNym,
            << "\n\n";
 }
 
-bool OTServerConnection::send(const OTEnvelope& theEnvelope)
+bool OTServerConnection::send(const String& theEnvelope)
```
Comment (mhamrle) [inline src/client/OTServerConnection.cpp]: `theEnvelope` is not envelope now, the variable should be renamed.

File: src/core/OTServerContract.cpp (hunk 1 above)
Comment (benma) [inline src/core/OTServerContract.cpp]: Fixed.

**Comment by benma** [comment]
> It's probably this one with the `_bin` suffix:

You were right. Fixed.

File: src/client/OTServerConnection.cpp (hunk 2 above)
Comment (benma) [inline src/client/OTServerConnection.cpp]: Fixed.

**Comment by benma** [comment]
test this please

**Comment by mhamrle** [comment]
:+1:

**Comment by weissjeffm** [comment]
test this please

File: deps/czmq [hunk 3]
```diff
@@ -1 +1 @@
-Subproject commit 869c3145b2ae60802f6239b9615b2bcd8b715d96
+Subproject commit 9c24aea3defe84cd72d7e6ac1845b6215f3d7aad
```
Comment (toxeus) [inline deps/czmq]: Please don't downgrade to the `v3.0.0` tag. I have checked out 869c3145b2ae60802f6239b9615b2bcd8b715d96 for a reason: the `CZMQ_BUILD_TESTS` variable has been introduced which we are [using to disable the czmq tests](https://github.com/Open-Transactions/opentxs/blob/develop/deps/CMakeLists.txt#L62-L63).

File: src/ext/CMakeLists.txt [hunk 4]
```diff
@@ -1,5 +1,11 @@
 # Copyright (c) Monetas AG, 2014
 
+include_directories(SYSTEM
+  ${ZEROMQ_INCLUDE_DIRS}
+  ${CZMQ_INCLUDE_DIR}
+)
+
+
```
Comment (toxeus) [inline src/ext/CMakeLists.txt]: Why is this necessary when you did not touch any files in `src/ext`?

File: include/opentxs/core/OTServerContract.hpp [hunk 5]
```diff
@@ -166,6 +169,7 @@ class OTServerContract : public Contract
     String m_strHostname;
     int32_t m_nPort;
     String m_strURL;
+    unsigned char* m_transportKey;
```
Comment (toxeus) [inline include/opentxs/core/OTServerContract.hpp]: I don't know if this type is forced by czmq but I'd prefer `uint8_t`.

**Comment by toxeus** [comment]
@monetas-ci re-test this please

File: deps/czmq (hunk 3 above)
Comment (benma) [inline deps/czmq]: That was not deliberate. I don't know what's wrong with this submodule, they always show up as changed in `git status` (along with another submodule). Will fix tomorrow.

File: include/opentxs/core/OTServerContract.hpp (hunk 5 above)
Comment (benma) [inline include/opentxs/core/OTServerContract.hpp]: "forced" is the wrong word in a language where you can do reinterpret_cast ;) Anyway, the czmq functions do work with `unsigned char*`, and I don't mind it because this is the convention for binary data (as opposed to `char*` for strings). If you insist on `uint8_t`, I can change it, but I see no reason why it is better. Why do you prefer it?

File: src/ext/CMakeLists.txt (hunk 4 above)
Comment (benma) [inline src/ext/CMakeLists.txt]: If I remember correctly, OTPayment depends on one of the core/* classes where I include czmq.h. I will double-check tomorrow.

Comment (benma) [inline src/ext/CMakeLists.txt]: OTPayment.hpp -> InstantiateContract.hpp -> OTServerContract -> czmq.h.

File: src/core/CMakeLists.txt [hunk 6]
```diff
@@ -118,8 +123,8 @@ else()
   )
 endif()
 
-target_link_libraries(opentxs-core PRIVATE opentxs-recurring opentxs-script opentxs-cron opentxs-trade otprotob irrxml bitcoin-base58)
-target_link_libraries(opentxs-core PUBLIC ${PROTOBUF_LIBRARY} ${OPENSSL_LIBRARIES} ${ZLIB_LIBRARY} ${OPENTXS_SYSTEM_LIBRARIES} ${keyring})
+target_link_libraries(opentxs-core PRIVATE opentxs-recurring opentxs-script opentxs-cron opentxs-trade otprotob irrxml bitcoin-base58 czmq_local)
```
Comment (toxeus) [inline src/core/CMakeLists.txt]: You don't need to put `czmq_local` also here.

File: include/opentxs/core/OTServerContract.hpp (hunk 5 above)
Comment (toxeus) [inline include/opentxs/core/OTServerContract.hpp]: Let's not use `reinterpret_cast`. Just wanted to check if it's czmq that uses that convention. [More info](http://stackoverflow.com/questions/16138237/when-is-uint8-t-%E2%89%A0-unsigned-char/16138470#16138470)

File: src/core/CMakeLists.txt (hunk 6 above)
Comment (benma) [inline src/core/CMakeLists.txt]: Where? It is there, in both `target_link_libraries` statements.

**Comment by toxeus** [comment]
@monetas-ci re-test this please

**Comment by toxeus** [comment]
@monetas-ci re-test this please

**Comment by toxeus** [comment]
:+1: excellent
